File: packages/db/src/schema/job_.ts
// import {
//   jsonb,
//   pgEnum,
//   pgTable,
//   text,
//   timestamp,
//   uuid,
// } from "drizzle-orm/pg-core";

// import { environment } from "./environment.js";
// import { release } from "./release.js";
// import { target } from "./target.js";

// export const jobAgentConfig = pgTable("job_agent_config", {
//   id: uuid("id").primaryKey().defaultRandom().notNull(),
//   jobId: uuid("job_id"),
//   agentId: uuid("job"),
//   config: jsonb("config").default("{}"),
// });

// export const releaseJobTrigger = pgTable("release_job_trigger", {
//   id: uuid("id").primaryKey().defaultRandom().notNull(),

//   targetId: uuid("target_id")
//     .references(() => target.id)
//     .notNull(),
//   releaseId: uuid("release_id")
//     .references(() => release.id)
//     .notNull(),
//   environmentId: uuid("environment_id")
//     .references(() => environment.id)
//     .notNull(),

//   jobId: uuid("job_id")
//     .references(() => job.id)
//     .unique(),
// });

// export const jobStatus = pgEnum("job_status", [
//   "completed",
//   "cancelled",
//   "skipped",
//   "in_progress",
//   "action_required",
//   "pending",
//   "failure",
//   "invalid_job_agent",
//   "invalid_integration",
//   "external_run_not_found",
// ]);

// export const jobReason = pgEnum("job_reason", [
//   "policy_passing",
//   "policy_override",
//   "env_policy_override",
//   "config_policy_override",
// ]);

// export const job = pgTable("job", {
//   id: uuid("id").primaryKey().defaultRandom().notNull(),

//   jobAgentId: uuid("job_agent_id")
//     .notNull()
//     .references(() => jobAgent.id)
//     .notNull(),

//   jobAgentConfig: jsonb("job_agent_config")
//     .notNull()
//     .default("{}")
//     .$type<Record<string, any>>(),

//   metadata: jsonb("metadata").default("{}").$type<Record<string, any>>(),
//   status: jobStatus("status").notNull().default("pending"),
//   message: text("message"),
//   reason: jobReason("reason").notNull().default("policy_passing"),
//   createdAt: timestamp("created_at").defaultNow(),
//   updatedAt: timestamp("updated_at")
//     .defaultNow()
//     .$onUpdate(() => new Date()),
// });

File: apps/workspace-engine/pkg/model/job/job.go
package job

import "time"

type JobStatus string

const (
	JobStatusPending             JobStatus = "pending"
	JobStatusSkipped             JobStatus = "skipped"
	JobStatusInProgress          JobStatus = "in_progress"
	JobStatusActionRequired      JobStatus = "action_required"
	JobStatusCancelled           JobStatus = "cancelled"
	JobStatusFailure             JobStatus = "failure"
	JobStatusInvalidJobAgent     JobStatus = "invalid_job_agent"
	JobStatusInvalidIntegration  JobStatus = "invalid_integration"
	JobStatusExternalRunNotFound JobStatus = "external_run_not_found"
	JobStatusSuccessful          JobStatus = "successful"
)

type JobReason string

const (
	JobReasonPolicyPassing        JobReason = "policy_passing"
	JobReasonPolicyOverride       JobReason = "policy_override"
	JobReasonEnvPolicyOverride    JobReason = "env_policy_override"
	JobReasonConfigPolicyOverride JobReason = "config_policy_override"
)

type Job struct {
	ID string `json:"id"`

	JobAgentID     *string        `json:"jobAgentId,omitempty"`
	JobAgentConfig map[string]any `json:"jobAgentConfig"`

	ExternalID *string `json:"externalId,omitempty"`

	Status  JobStatus `json:"status"`
	Reason  JobReason `json:"reason"`
	Message *string   `json:"message,omitempty"`

	CreatedAt   time.Time `json:"createdAt"`
	StartedAt   time.Time `json:"startedAt,omitempty"`
	CompletedAt time.Time `json:"completedAt,omitempty"`
	UpdatedAt   time.Time `json:"updatedAt"`
}

func (j Job) GetID() string {
	return j.ID
}

func (j Job) GetStatus() JobStatus {
	return j.Status
}

func (j Job) GetReason() JobReason {
	return j.Reason
}

func (j Job) GetMessage() *string {
	return j.Message
}

func (j Job) GetCreatedAt() time.Time {

	return j.CreatedAt
}

func (j Job) GetStartedAt() time.Time {
	return j.StartedAt
}

func (j Job) GetCompletedAt() time.Time {
	return j.CompletedAt
}

func (j Job) GetUpdatedAt() time.Time {
	return j.UpdatedAt
}

func (j Job) GetJobAgentID() *string {
	return j.JobAgentID
}

func (j Job) GetJobAgentConfig() map[string]any {
	return j.JobAgentConfig
}

func (j Job) GetExternalID() *string {
	return j.ExternalID
}

func (j Job) IsInProgress() bool {
	return j.Status == JobStatusInProgress || j.Status == JobStatusActionRequired
}

func (j Job) IsPending() bool {
	return j.Status == JobStatusPending
}

func (j Job) IsCompleted() bool {
	return j.Status == JobStatusSuccessful ||
		j.Status == JobStatusFailure ||
		j.Status == JobStatusCancelled ||
		j.Status == JobStatusSkipped ||
		j.Status == JobStatusInvalidJobAgent ||
		j.Status == JobStatusInvalidIntegration ||
		j.Status == JobStatusExternalRunNotFound
}

func (j Job) IsSuccessful() bool {
	return j.Status == JobStatusSuccessful
}

func (j Job) IsFailure() bool {
	return j.Status == JobStatusFailure ||
		j.Status == JobStatusExternalRunNotFound ||
		j.Status == JobStatusInvalidJobAgent ||
		j.Status == JobStatusInvalidIntegration
}
